### DISCUSSION TITLE
Is it possible to have options in nodes as there is for ContextSwitcher ?

### DISCUSSION THREAD

**Discussion by luddwskater** [SAP/luigi#2952] [General]
I need more advanced navigation as:
1) top menu with options ( as ContexSwitcher but many of them in top bar )
2) context switcher acting on Microfrontend event ( left menu added when one record is presented from a list ).

Have you an idea how to get this ?

**Comment by ndricimrr** [comment]
Hi @luddwskater
> 1. top menu with options ( as ContexSwitcher but many of them in top bar )

If I get this right you want to be able to add multiple **ContextSwitcher**-like dropdowns or similar items at the top navigation bar for more advances scenarios. This cannot be achieved with a simple _Luigi Configuration_ change as the top navigation elements are fixed to a specific navigation bar design that we adhere to. 
However, should you want to further change the top navigation bar to fit your needs you can do so by removing the top navigation bar and adding your own, based on your specific needs and designs. There the [Luigi Core API](https://docs.luigi-project.io/docs/luigi-core-api) may come in handy to simulate the navigation links and much more.
To get you started you can have a look at the FAQ which gives also a small example of that :

[I don't want to use the default Fiori Fundamentals style. How can I style Luigi differently? How can I change the look of Luigi's UI?
Can I hide the Luigi default side/top navigation panel?](https://docs.luigi-project.io/docs/faq?section=ui-questions)


> 2. context switcher acting on Microfrontend event ( left menu added when one record is presented from a list ).

If I understand this correctly, you are trying to have an event in a microfrontend affect let's say the **Context Switcher** elements only. I would say such a scenario should be possible through using one or a combination of the Luigi Core API calls like : [Luigi.setConfig()](https://docs.luigi-project.io/docs/luigi-core-api/?section=setconfig),  [Luigi.getConfig()](https://docs.luigi-project.io/docs/luigi-core-api/?section=getconfig) and/or [Luigi.configChanged()](https://docs.luigi-project.io/docs/luigi-core-api/?section=configchanged) to change the Context Switcher configurations.

However a microfrontend is not allowed to inflict a direct change to the Luigi configuration for obvious safety reasons. Instead you can communicate a message request from the client microfrontend to the core and handle the Context Switcher config change on the core side. 
One small example of how this would look like in pseudocode that I could think of is:

Have a look at the [Client->Core communcation here](https://docs.luigi-project.io/docs/communication)
// your luigi-config.js file would look like this

```
 communication: {
    customMessagesListeners: {
      'my-custom-message.change-top-nav-context-switcher': (message) => {
          // Change the config to add other options
          Luigi.getConfig().navigation.contextSwitcher.options.push({
             label: message.newData.label,
             pathValue: message.newData.path
          });
          // Tell Luigi the context switcher config has been changed
          Luigi.configChanged('navigation.contextSwitcher');
      }
    }
  }
```

and in the client microfrontend you would have something like this:

```
addContextSwitcherItem(){
   LuigiClient.sendCustomMessage({id: 'my-custom-message.change-top-nav-context-switcher', newData: {
      label: 'New Label', 
      path: 'my-new-path'
   }})
}
```

Let us know if you have any further questions.
